pbc additions + multi-subensemble

timothy-wohlfromm opened the pull request.

timothy-wohlfromm committed:
Update correlation.py

multi-subensemble, backward compatible

timothy-wohlfromm committed:
Update pbc.py

pbc_diff changed for arbitrary triclinic boxes

nielsmde commented on mdevaluate/correlation.py:
```diff
@@ -15,20 +15,56 @@ def log_indices(first, last, num=100):
+                if is_first_frame_loop:
+                    count[index]    = len(sf_sel)
+                f_values[index] = function(sf_sel, act_frame[sel]) if count[index] != 0 else 0
+            is_first_frame_loop = False
```
Kann es sein, dass `is_first_frame=False` eingerückt sein sollte?

timothy-wohlfromm replied:
Nein, aber so wie es da steht hat es keine Funktion - hatte das erst mit nem array gelöst (um Probleme mit dem scope zu umgehen) und dann das Falsch geändert.
Ich pass es an.

nielsmde commented on mdevaluate/correlation.py:
```diff
@@ -77,31 +121,42 @@ def shifted_correlation(function, frames,
+        np.divide(result, count[...,np.newaxis], out = result, where = count[...,np.newaxis] != 0)
+        count  = count  / len(start_frames)
+    else:
+        result = np.asarray(result)
```
Sehe ich das richtig, wenn `average=False` ist, ist `result` noch mit den counts multipliziert?

timothy-wohlfromm replied:
Wenn average = False, dann kommt der nicht in Zeile 146, sondern in 149 - müsste passen.

ghost commented:
Ich habe grade mal drüber gelesen und ein paar Kommentare/Frage in den Code geschrieben.

Ein großes Problem sehe ich momentan mit der jetztigen Umsetzung:
Wenn sich die Funktion `correlation` ändert, da wird ja jetzt auch der count mit zurück gegeben, ändert sich die Checksum der Funktionsaufrufe, was dazu führt, dass alle per autosave gespeicherten Auswertungen neu berechnet werden.

Wäre es nicht möglich, die Funktion `correlation` beim Alten zu belassen und in `shifted_correlation` zu unterscheiden, ob es counts gibt oder nicht? Dafür hast du ja auch schon das keyword Argument `counts=False` eingeführt.

timothy-wohlfromm committed:
Update correlation.py

scope von is_first_frame_loop erweitert - damit es funktioniert. So wird nur einmal pro Subset und Fenster die Länge bestimmt.

timothy-wohlfromm commented:
Wann ändert sich die checksum? Nur wenn sich der letzte output ändert?
Dann hätt ich nen Fix, der bei counter=False auch bei shifted_correlation die alten Ergebnisse liefert (Mittel ohne Gewichtung).

timothy-wohlfromm committed:
Update correlation.py

kwarg counter entfernt - alten correlationfcts wieder hergestellt, multi_subensemble_correlation eingefügt, innere funktion mit decorator versehen, kommentar in halbwegs verständlich geändert und noch ne abfrage auf sliceobjekt eingefügt(ist slice nicht in einer liste, gabs nen fehler).

timothy-wohlfromm committed:
Update pbc.py

pbc_diff changed

Commit:
Check if box actually has 'shape' attribute.

Commit:
Uses getattr with default value

I just realized that we can use getattr with a default value to safely retrieve attributes.

Commit:
Use getattr wuth default value

timothy-wohlfromm committed:
fix:  arbitrary function output

if average ==False: empty list -> append results -> make to array 
if average and counter ==False:  mean over 0th axis ( same as old shifted_corr)
if average and counter ==True: 
get number of dimensions for counter and result.
the first number in shape is #segments in both arrays !
the second number in shape is #of times in result -> add newaxis (line 167)
the next numbers in shape are same for both arrays (shape of selector) (line 172)
all following numbers in result (if any) will be shape of function output -> add newaxes (line 170)
build tupel for broadcasting (bt)
@line174: bt[1:] --- because of summation over axis=0 (dim reduction)
line175: moveaxis (times) to correct position -> ( (shape of sel) + #times, + (shape of fct output))
line 178: in case average is false the correct position is shifted by 1

line67 - allow for arbitrary function-output-shape

nielsmde commented on mdevaluate/correlation.py:
```diff
@@ -149,38 +149,41 @@ def correlate(start_frame):
-            else:
-                result.append(act_result)
+            act_result = list(correlate(start_frame))
+            result.append(act_result)
```
Wieso hast du denn hier geändert, dass bei average=True nicht mehr addiert wird? Das ist wichtig um Speicher zu sparen...

timothy-wohlfromm committed:
Update correlation.py

sum on the fly

Commit:
Change name of adhoc function that is used for autosave filename
